### DISCUSSION TITLE
Dev Session 6 May

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#1632] [General]
When: 6 May --- 2000 UTC+8/1400 CET/1200 GMT/0800 EDT

Where: [via Matrix](https://matrix.to/#/#mesa-dev:matrix.org)

Proposed Agenda:

- PyCon Sprints
- Mesa-Geo
- GSOC
- Mesa 2.0
 - Build Poll Results

@jackiekazil @rht @wang-boyu @Tortar @EwoutH Please feel free to add or takeaway from this agenda

**Comment by rht** [comment]
Adding:
- #1625 for replacing Matplotlib with Seaborn for the tutorial. Grammar of graphics definitely makes for a more comprehensible plotting code.
- Various implementations of `get_heading` in `ContinuousSpace`: #1620 (by @EwoutH ), #1664 (by @catherinedevlin), and current main. #1620 has the shortest norm overall, when I did a grid scan. 3 of them are equivalent when applied with modulo, as @jeremander said. Neither preserves the angle of the vector, due to wrapping. However, I'm wary of #1620, because `heading + self.half_size` changes the heading angle that is not due to wrapping. I'm more inclined to prefer #1664, if it can be made more performant.

**Reply by wang-boyu** [reply]
Speaking of grammar of graphics, Seaborn has a new `seaborn.objects` interface that is under development: [The seaborn.objects interface](https://seaborn.pydata.org/tutorial/objects_interface.html)

Another visualization library that I use often these days is `lets-plot` from JetBrains: https://lets-plot.org

**Comment by tpike3** [comment]
### Dev Notes

**PyCon Sprints**
-- Fixed bugs
-- Started and continuing to work on an overhaul of our documentation 
-- Cleared out a few stale PRs that were hanging out 
-- Made some of our commands more user friendly
-- Onboarded a few contributors to the project
-- Got a few people excited to learn about Mesa and agent-based modeling 
-- [Found a gnarly bug in our CI/CD pipeline in Codespell that doesn't handle images or pngs well that was produced by Jupyter notebooks and does not allow for regexes either to ignore them.](https://github.com/projectmesa/mesa/issues/1661) [This ticket was submitted to try to solve the Jupyter notebook parsing issue with codespell](https://github.com/codespell-project/codespell/issues/2851)

**GSoC**

Unfortunately Google was only able to provide 25 slots and had NUMFOCUS had 43 Tier 1s. To account for this NUMFOUCS randomly selected the 25 of the submitted Tier 1s and Mesa was not selected. 

To still try and get a Python based Mesa frontend built, we are going to submit for a Small Business Development Grant via NUMFOCUS.


**Development**

@glnnlhmn did some great work on the tutorial and this prompted some discussion about Mesa's tutorial approach. At the June dev meeting we will discuss a comprehensive approach to Mesa's tutorials and documentation using resources such as [Diataxis](https://diataxis.fr/) 

@rht is going to update #1664 so it calculates the norm from scratch which will make it more performant. 

@jackiekazil is going to look through the [Mesa 2.0](https://github.com/projectmesa/mesa/milestone/33) milestone tickets and move those that are not breaking changes to [Next release](https://github.com/projectmesa/mesa/milestone/36), to determine if we should do another release before Mesa. 2.0 

Poll results are in a small majority voted for cython improvements to be part of a separate build and not the standard Mesa build. 

**CSSSA**

- @tpike3 is going to submit to do a tutorial for Mesa and Mesa-Geo at the annual meeting of [CSSSA ](https://computationalsocialscience.org/conferences/css2023/)

**Reply by rht** [reply]
I asked GPT-3.5 about the heading in the `ContinuousSpace`. Got this answer
```python
import numpy as np

# Define the two vectors
vector1 = np.array([5, 10, 15])
vector2 = np.array([3, 12, 20])

# Specify the size of the toroidal space
size = np.array([7, 7, 7])

# Calculate the wrap-around differences in each dimension
diffs = []
for dim in range(len(vector1)):
    diff = abs(vector1[dim] - vector2[dim])
    wrap_diff = min(diff, size[dim] - diff)
    diffs.append(wrap_diff)

# Calculate the smallest difference
smallest_diff = np.sqrt(np.sum(np.array(diffs) ** 2))

print(smallest_diff)
```
1. This makes sense, as the diff for each dimension can and should be independently optimized. Existing solutions haven't implemented this
2. Still wrong because the diff for each dimension is calculated as their magnitude, which has no info on directionality of the diff

I will modify GPT-3.5's solution, fixing for point 2. Not sure how the attribution works. At least I have given attribute in this very post.

**Reply by ameligrana** [reply]
Didn't see this discussion before but I wrote a performant version of this function in Agents some time ago:

```julia
function get_direction(
    from::ValidPos,
    to::ValidPos,
    space::Union{ContinuousSpace{D,true},AbstractGridSpace{D,true}},
) where {D}
    direct_dir = to .- from
    inverse_dir = direct_dir .- sign.(direct_dir) .* spacesize(space)
    return map((x, y) -> argmin(abs, (x, y)), direct_dir, inverse_dir)
end
```

the same with more verbosity can be done in Python, hope it helps!

**Reply by rht** [reply]
I suppose it can't be concise without NumPy, but in the current implementation, the diff is already calculated via NumPy anyway.

**Reply by rht** [reply]
Made #1686 for a less wrong solution.
